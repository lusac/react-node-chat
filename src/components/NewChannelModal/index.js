import React from 'react'
import PropTypes from 'prop-types'

export default class NewChannelModal extends React.Component {
  constructor() {
    super()
    this.state = {
      channelName: ''
    }
  }

  createChat(channelName) {
    if (this.props.socket) {
      this.props.socket.emit('create channel', channelName)
    }
  }

  onChatNameChange(e) {
    this.setState({
      channelName: e.target.value
    })
  }

  onFormSubmit(e) {
    e.preventDefault()
    if (this.state.channelName) {
      this.createChat(this.state.channelName)
      this.props.toggleModal()
    }
  }

  render() {
    return (
      <div className="modal">
        <div className="modal__container">
          <div className="modal__dialog">
            <form onSubmit={this.onFormSubmit.bind(this)}>
              <span
                className="modal__close"
                onClick={this.props.toggleModal.bind(this)}>fechar</span>

              <h3>Crie um canal</h3>

              <label>nome</label>
              <input
                type="text"
                placeholder="ex: canal da galera"
                onChange={this.onChatNameChange.bind(this)}></input>

              <span className="modal__dialog__btns">
                <button
                  className="red"
                  type="button"
                  onClick={this.props.toggleModal.bind(this)}>cancelar</button>
                <button
                  disabled={!this.state.channelName}
                  className="blue"
                  type="submit">salvar</button>
              </span>
            </form>
          </div>
        </div>
      </div>
    )
  }
}

NewChannelModal.propTypes = {
  socket: PropTypes.object,
  toggleModal: PropTypes.func
};