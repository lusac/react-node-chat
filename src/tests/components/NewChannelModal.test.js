/* global it describe expect beforeEach jest */
import React from 'react'
import { shallow, configure } from 'enzyme'
import renderer from 'react-test-renderer'
import Adapter from 'enzyme-adapter-react-16'
import NewChannelModal from '../../components/NewChannelModal'

configure({ adapter: new Adapter() })

describe('NewChannelModal component', () => {
  describe('snapshot', () => {
    it('default', () => {
      const
        renderedValue =  renderer.create(<NewChannelModal toggleModal={() => 'hello!' } />).toJSON()
      expect(renderedValue).toMatchSnapshot()
    })
  })

  describe('methods', () => {
    let
      wrapper,
      socket = {
        emit: (a, b) => (a, b)
      }

    beforeEach(() => {
      wrapper = shallow(<NewChannelModal socket={socket} toggleModal={() => 'Hello!' } />)
    })

    it('createChat method', () => {
      const spy = jest.spyOn(wrapper.instance().props.socket, 'emit')
      wrapper.instance().createChat('My Channel')
      expect(spy).toHaveBeenCalledWith('create channel', 'My Channel')
    })

    it('onChatNameChange method', () => {
      expect(wrapper.instance().state.channelName).toBe('')
      wrapper.instance().onChatNameChange({target: {value: 'new name'}})
      expect(wrapper.instance().state.channelName).toBe('new name')
    })

    it('onFormSubmit method', () => {
      const e = { preventDefault: () => 'oie' }
      wrapper.instance().onChatNameChange({target: {value: 'new name'}})

      const spyPreventDefault = jest.spyOn(e, 'preventDefault')
      const spyCreateChat = jest.spyOn(wrapper.instance(), 'createChat')
      // const spyToggleModal = jest.spyOn(wrapper.instance().props, 'toggleModal')

      wrapper.instance().onFormSubmit(e)

      expect(spyPreventDefault).toHaveBeenCalled()
      expect(spyCreateChat).toHaveBeenCalledWith(wrapper.instance().state.channelName)
      // expect(spyToggleModal).toHaveBeenCalled()
    })
  })
})