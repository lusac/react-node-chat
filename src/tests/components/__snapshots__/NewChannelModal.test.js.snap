// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewChannelModal component snapshot default 1`] = `
<div
  className="modal"
>
  <div
    className="modal__container"
  >
    <div
      className="modal__dialog"
    >
      <form
        onSubmit={[Function]}
      >
        <span
          className="modal__close"
          onClick={[Function]}
        >
          âœ•
        </span>
        <h3>
          Crie um canal
        </h3>
        <label>
          nome
        </label>
        <input
          onChange={[Function]}
          placeholder="ex: canal da galera"
          type="text"
        />
        <span
          className="modal__dialog__btns"
        >
          <button
            className="red-transparent"
            onClick={[Function]}
            type="button"
          >
            cancelar
          </button>
          <button
            className="blue"
            disabled={true}
            type="submit"
          >
            salvar
          </button>
        </span>
      </form>
    </div>
  </div>
</div>
`;
